File: mercury/transitive_group.m
%---------------------------------------------------------------------------%
% vim: ft=mercury ts=4 sw=4 et
%---------------------------------------------------------------------------%
% File: transitive_group.m
% Main author: nyc
%
% Database and runtime computation of transitive subgroups of Sₙ.
%
% Transitive subgroups of Sₙ (up to conjugacy) classify the possible
% Galois groups of irreducible degree-n polynomials over Q.
%
% For prime p, the solvable transitive subgroups of Sₚ are exactly
% Z/p ⋊ H where H ≤ (Z/pZ)*, one per divisor d of p-1. These are
% computed at runtime from the structure of AGL(1,p).
%
% Groups are numbered following the Butler-McKay convention (1983),
% which is also the standard in GAP, Magma, and the LMFDB.
%
%---------------------------------------------------------------------------%

:- module transitive_group.
:- interface.

:- import_module bool.
:- import_module integer.
:- import_module list.
:- import_module maybe.
:- import_module permutation.

%---------------------------------------------------------------------------%

:- type transitive_group
    --->    transitive_group(
                tg_name                :: string,
                tg_degree              :: int,
                tg_order               :: integer,
                tg_generators          :: list(perm),
                tg_solvable            :: bool,
                tg_maximal_supergroups :: list(int),
                tg_composition_factors :: list(int)
            ).

    % All transitive subgroups of Sₙ for the given degree.
    % Covers degree 5 (hard-coded) and all prime degrees >= 3
    % (computed at runtime from AGL(1,p) structure).
    %
:- func trans_groups_of_degree(int) = list(transitive_group).

    % Find transitive group(s) of given degree and order.
    %
:- func trans_group_by_order(int, integer) = list(transitive_group).

    % Test whether a transitive group is solvable.
    %
:- func is_solvable(transitive_group) = bool.

    % For a solvable group, return the composition series as a list
    % of generating sets, descending from G to {1}.
    % Returns no for non-solvable groups.
    %
:- func composition_series(transitive_group)
    = maybe(list(list(perm))).

%---------------------------------------------------------------------------%
%---------------------------------------------------------------------------%

:- implementation.

:- import_module int.
:- import_module positive.
:- import_module prime_factors.
:- import_module set.
:- import_module string.

%---------------------------------------------------------------------------%

trans_groups_of_degree(N) = Result :-
    ( if N = 5 then
        Result = degree5_groups
    else if int.'>='(N, 3), is_prime(integer(N)) then
        Result = trans_groups_of_prime_rt(N)
    else
        Result = []
    ).

trans_group_by_order(Deg, Ord) =
    list.filter(
        ( pred(G::in) is semidet :- tg_order(G) = Ord ),
        trans_groups_of_degree(Deg)).

is_solvable(G) = tg_solvable(G).

composition_series(TG) = Result :-
    ( if tg_solvable(TG) = no then
        Result = no
    else
        Name = tg_name(TG),
        ( if Name = "C5" then
            Result = yes([tg_generators(TG), []])
        else if Name = "D5" then
            Result = yes([
                tg_generators(TG),
                [from_cycles(5, [[0, 1, 2, 3, 4]])],
                []])
        else if Name = "F20" then
            Result = yes([
                tg_generators(TG),
                [from_cycles(5, [[0, 1, 2, 3, 4]]),
                 from_cycles(5, [[1, 4], [2, 3]])],
                [from_cycles(5, [[0, 1, 2, 3, 4]])],
                []])
        else if is_prime(integer(tg_degree(TG))) then
            Result = composition_series_prime(TG)
        else
            Result = no
        )
    ).

%---------------------------------------------------------------------------%
% Degree 5 transitive groups (hard-coded fast path)
%---------------------------------------------------------------------------%

:- func degree5_groups = list(transitive_group).

degree5_groups = [
    % T1: C5 = ⟨(0 1 2 3 4)⟩
    transitive_group(
        "C5", 5, integer(5),
        [from_cycles(5, [[0, 1, 2, 3, 4]])],
        yes,
        [1],
        [5]),

    % T2: D5 = ⟨(0 1 2 3 4), (1 4)(2 3)⟩
    transitive_group(
        "D5", 5, integer(10),
        [from_cycles(5, [[0, 1, 2, 3, 4]]),
         from_cycles(5, [[1, 4], [2, 3]])],
        yes,
        [2],
        [5, 2]),

    % T3: F20 = ⟨(0 1 2 3 4), (1 2 4 3)⟩
    transitive_group(
        "F20", 5, integer(20),
        [from_cycles(5, [[0, 1, 2, 3, 4]]),
         from_cycles(5, [[1, 2, 4, 3]])],
        yes,
        [3, 4],
        [5, 2, 2]),

    % T4: A5 = ⟨(0 1 2 3 4), (0 1 2)⟩
    transitive_group(
        "A5", 5, integer(60),
        [from_cycles(5, [[0, 1, 2, 3, 4]]),
         from_cycles(5, [[0, 1, 2]])],
        no,
        [4],
        []),

    % T5: S5 = ⟨(0 1 2 3 4), (0 1)⟩
    transitive_group(
        "S5", 5, integer(120),
        [from_cycles(5, [[0, 1, 2, 3, 4]]),
         from_cycles(5, [[0, 1]])],
        no,
        [],
        [])
].

%---------------------------------------------------------------------------%
% Runtime AGL(1,p) computation for prime degrees
%---------------------------------------------------------------------------%

    % Modular exponentiation: mod_exp(B, E, M) = B^E mod M.
    %
:- func mod_exp_rt(integer, integer, integer) = integer.

mod_exp_rt(B, E, M) = Result :-
    ( if integer.is_zero(E) then
        Result = integer.one
    else if integer.'mod'(E, integer(2)) = integer.zero then
        Half = mod_exp_rt(B, E // integer(2), M),
        Result = integer.'mod'(Half * Half, M)
    else
        Result = integer.'mod'(B * mod_exp_rt(B, E - integer.one, M), M)
    ).

    % Find a primitive root modulo a prime p via trial.
    %
:- func primitive_root_rt(integer) = integer.

primitive_root_rt(P) = Result :-
    Phi = P - integer.one,
    Factors = prime_factors(det_positive(Phi)),
    ( if find_primitive_root(P, Phi, Factors, integer(2), G) then
        Result = G
    else
        Result = integer(2)  % should not happen for primes
    ).

:- pred find_primitive_root(integer::in, integer::in, list(integer)::in,
    integer::in, integer::out) is semidet.

find_primitive_root(P, Phi, Factors, G, Result) :-
    integer.'<'(G, P),
    ( if is_prim_root(G, P, Phi, Factors) then
        Result = G
    else
        find_primitive_root(P, Phi, Factors, G + integer.one, Result)
    ).

:- pred is_prim_root(integer::in, integer::in, integer::in,
    list(integer)::in) is semidet.

is_prim_root(_, _, _, []).
is_prim_root(G, P, Phi, [Q | Qs]) :-
    mod_exp_rt(G, Phi // Q, P) \= integer.one,
    is_prim_root(G, P, Phi, Qs).

    % Sorted positive divisors of n.
    %
:- func divisors_rt(integer) = list(integer).

divisors_rt(N) = Result :-
    integer_sqrt_rt(N, Sq),
    collect_divisors(N, integer.one, Sq, [], Divs),
    list.sort_and_remove_dups(Divs, Result).

:- pred collect_divisors(integer::in, integer::in, integer::in,
    list(integer)::in, list(integer)::out) is det.

collect_divisors(N, D, Sq, Acc, Result) :-
    ( if integer.'>'(D, Sq) then
        Result = Acc
    else if integer.'mod'(N, D) = integer.zero then
        Q = N // D,
        ( if D = Q then
            collect_divisors(N, D + integer.one, Sq, [D | Acc], Result)
        else
            collect_divisors(N, D + integer.one, Sq, [D, Q | Acc], Result)
        )
    else
        collect_divisors(N, D + integer.one, Sq, Acc, Result)
    ).

:- pred integer_sqrt_rt(integer::in, integer::out) is det.

integer_sqrt_rt(N, Result) :-
    ( if integer.'=<'(N, integer.zero) then
        Result = integer.zero
    else
        % Newton's method from a rough starting point
        NF = integer.det_to_int(N),
        Approx0 = integer(float.truncate_to_int(
            math.sqrt(float.float(NF)))),
        Start = ( if integer.'<'(Approx0, integer.one) then
                      integer.one
                  else
                      Approx0 + integer.one ),
        isqrt_newton_rt(N, Start, Result)
    ).

:- pred isqrt_newton_rt(integer::in, integer::in, integer::out) is det.

isqrt_newton_rt(N, X, Result) :-
    XP = (X + N // X) // integer(2),
    ( if integer.'>='(XP, X) then
        Result = X
    else
        isqrt_newton_rt(N, XP, Result)
    ).

:- use_module float.
:- use_module math.

    % Prime factorisation as a flat list with multiplicity.
    % E.g. prime_factors_with_mult(12) = [2, 2, 3].
    %
:- func prime_factors_with_mult_rt(integer) = list(integer).

prime_factors_with_mult_rt(N) = Result :-
    ( if N = integer.one then
        Result = []
    else
        Fs = factorise(det_positive(N)),
        Result = list.condense(list.map(
            ( func(PP) = list.duplicate(pp_exponent(PP), pp_prime(PP)) ),
            Fs))
    ).

    % Build all transitive subgroups of Sₚ for a prime p.
    %
    % Solvable subgroups: Z/p ⋊ H for each divisor d of p-1.
    % Non-solvable: Aₚ and Sₚ.
    % Maximal supergroups are computed from divisibility.
    %
:- func trans_groups_of_prime_rt(int) = list(transitive_group).

trans_groups_of_prime_rt(N) = Result :-
    P = integer(N),
    G = primitive_root_rt(P),
    Ds = divisors_rt(P - integer.one),
    SolvableGroups = list.map(mk_affine_group(N, P, G), Ds),
    % Aₚ
    Factorial = factorial_integer(P),
    Ap = transitive_group(
        "A" ++ string.int_to_string(N),
        N,
        Factorial // integer(2),
        [from_cycles(N, [0 `..` int.'-'(N, 1)]),
         from_cycles(N, [[0, 1, 2]])],
        ( if int.'<'(N, 5) then yes else no ),
        [],
        []),
    % Sₚ
    Sp = transitive_group(
        "S" ++ string.int_to_string(N),
        N,
        Factorial,
        [from_cycles(N, [0 `..` int.'-'(N, 1)]),
         from_cycles(N, [[0, 1]])],
        ( if int.'<'(N, 4) then yes else no ),
        [],
        []),
    AllGroups0 = SolvableGroups ++ [Ap, Sp],
    list.sort(order_by_order, AllGroups0, AllGroupsSorted),
    % Assign maximal supergroups
    indexed_list(AllGroupsSorted, 0, Indexed),
    Result = list.map(assign_supergroups(Indexed), Indexed).

:- func factorial_integer(integer) = integer.

factorial_integer(N) = Result :-
    ( if integer.'=<'(N, integer.one) then
        Result = integer.one
    else
        Result = N * factorial_integer(N - integer.one)
    ).

:- pred order_by_order(transitive_group::in, transitive_group::in,
    comparison_result::out) is det.

order_by_order(A, B, Result) :-
    compare(Result, tg_order(A), tg_order(B)).

:- func mk_affine_group(int, integer, integer, integer) = transitive_group.

mk_affine_group(N, P, G, D) = Result :-
    Trans = from_mapping(list.map(
        ( func(I) = integer.det_to_int(
            integer.'mod'(integer(I) + integer.one, P)) ),
        0 `..` int.'-'(N, 1))),
    ScaleFactor = mod_exp_rt(G, (P - integer.one) // D, P),
    Scale = from_mapping(list.map(
        ( func(I) = integer.det_to_int(
            integer.'mod'(ScaleFactor * integer(I), P)) ),
        0 `..` int.'-'(N, 1))),
    Gens = ( if D = integer.one then [Trans] else [Trans, Scale] ),
    Name = affine_group_name(P, D),
    DFactors = prime_factors_with_mult_rt(D),
    CompFactors = list.map(integer.det_to_int, DFactors ++ [P]),
    Result = transitive_group(
        Name, N, P * D, Gens, yes, [], CompFactors).

:- func affine_group_name(integer, integer) = string.

affine_group_name(P, D) = Name :-
    PS = integer.to_string(P),
    DS = integer.to_string(D),
    ( if D = integer.one then
        Name = "Z" ++ PS
    else if D = integer(2) then
        Name = "D" ++ PS
    else if D = P - integer.one then
        Name = "AGL(1," ++ PS ++ ")"
    else
        Name = "Z" ++ PS ++ ":Z" ++ DS
    ).

:- pred indexed_list(list(T)::in, int::in,
    list({int, T})::out) is det.

indexed_list([], _, []).
indexed_list([X | Xs], I, [{I, X} | Rest]) :-
    indexed_list(Xs, int.'+'(I, 1), Rest).

:- func assign_supergroups(list({int, transitive_group}),
    {int, transitive_group}) = transitive_group.

assign_supergroups(AllIndexed, {MyIdx, TG}) = Result :-
    MyOrder = tg_order(TG),
    % Groups that could contain us (larger order, our order divides theirs)
    Candidates = list.filter(
        ( pred({I, CG}::in) is semidet :-
            I \= MyIdx,
            integer.'>'(tg_order(CG), MyOrder),
            integer.'mod'(tg_order(CG), MyOrder) = integer.zero ),
        AllIndexed),
    % A supergroup is maximal if no candidate has order strictly between
    MaxSupers = list.filter_map(
        ( func({I, SuperG}) = I is semidet :-
            SuperOrder = tg_order(SuperG),
            not has_intermediate(Candidates, MyOrder, SuperOrder)
        ),
        Candidates),
    Result = TG ^ tg_maximal_supergroups := MaxSupers.

    % Check whether any group in the candidate list has order strictly
    % between MyOrder and SuperOrder, dividing both.
    %
:- pred has_intermediate(list({int, transitive_group})::in,
    integer::in, integer::in) is semidet.

has_intermediate([{_, MidG} | Rest], MyOrder, SuperOrder) :-
    MidOrder = tg_order(MidG),
    ( if
        integer.'>'(MidOrder, MyOrder),
        integer.'<'(MidOrder, SuperOrder),
        integer.'mod'(SuperOrder, MidOrder) = integer.zero,
        integer.'mod'(MidOrder, MyOrder) = integer.zero
    then
        true
    else
        has_intermediate(Rest, MyOrder, SuperOrder)
    ).

%---------------------------------------------------------------------------%
% Composition series for prime-degree affine subgroups
%---------------------------------------------------------------------------%

    % Composition series for Z/p ⋊ H where H has order d.
    % Descends through subgroups of H by removing one prime factor at a time.
    %
:- func composition_series_prime(transitive_group)
    = maybe(list(list(perm))).

composition_series_prime(TG) = Result :-
    P = integer(tg_degree(TG)),
    N = tg_degree(TG),
    G = primitive_root_rt(P),
    D = tg_order(TG) // P,
    DFactors = prime_factors_with_mult_rt(D),
    % Chain of divisors: d, d/q1, d/(q1*q2), ..., 1
    DChain = scan_div(D, DFactors),
    Trans = from_cycles(N, [0 `..` int.'-'(N, 1)]),
    GensList = list.map(mk_gens_for_d(N, P, G, Trans), DChain),
    Result = yes(GensList ++ [[]]).

:- func scan_div(integer, list(integer)) = list(integer).

scan_div(Acc, []) = [Acc].
scan_div(Acc, [Q | Qs]) = [Acc | scan_div(Acc // Q, Qs)].

:- func mk_gens_for_d(int, integer, integer, perm, integer) = list(perm).

mk_gens_for_d(N, P, G, Trans, D) = Gens :-
    ( if integer.'=<'(D, integer.one) then
        Gens = [Trans]
    else
        SF = mod_exp_rt(G, (P - integer.one) // D, P),
        Scale = from_mapping(list.map(
            ( func(I) = integer.det_to_int(
                integer.'mod'(SF * integer(I), P)) ),
            0 `..` int.'-'(N, 1))),
        Gens = [Trans, Scale]
    ).

%---------------------------------------------------------------------------%
:- end_module transitive_group.
%---------------------------------------------------------------------------%

File: scala3/src/main/scala/surd/TransitiveGroup.scala
package surd

/** Database of transitive subgroups of S_n for small n.
  *
  * For Galois group identification, we need the lattice of transitive
  * subgroups with generators, orders, solvability flags, and
  * maximal-supergroup relationships.
  *
  * Numbering follows the Butler-McKay convention (also used in GAP/Magma/LMFDB).
  *
  * For prime p, transitive subgroups are computed at runtime from the
  * structure of AGL(1,p). The solvable subgroups are Z/p ⋊ H for each
  * divisor d of p-1, plus A_p and S_p (non-solvable for p >= 5).
  */
object TransitiveGroup:

  /** A transitive group record. */
  final case class TGroup(
      degree: Int,
      index: Int,
      name: String,
      order: Long,
      solvable: Boolean,
      generators: Vector[Perm],
      maximalSupergroups: Vector[Int] = Vector.empty,
      compositionFactors: Vector[Int] = Vector.empty
  )

  // --- Degree 5 transitive groups (hard-coded fast path) ---

  /** C5 (T1): cyclic group of order 5. */
  val c5: TGroup = TGroup(
    degree = 5, index = 1, name = "C5", order = 5, solvable = true,
    generators = Vector(Perm.cycle(5, 0, 1, 2, 3, 4)),
    maximalSupergroups = Vector(1),
    compositionFactors = Vector(5)
  )

  /** D5 (T2): dihedral group of order 10. */
  val d5: TGroup = TGroup(
    degree = 5, index = 2, name = "D5", order = 10, solvable = true,
    generators = Vector(
      Perm.cycle(5, 0, 1, 2, 3, 4),
      Perm(Vector(0, 4, 3, 2, 1)) // reflection
    ),
    maximalSupergroups = Vector(2),
    compositionFactors = Vector(5, 2)
  )

  /** F20 (T3): Frobenius group of order 20. Z/5 ⋊ Z/4. */
  val f20: TGroup = TGroup(
    degree = 5, index = 3, name = "F20", order = 20, solvable = true,
    generators = Vector(
      Perm.cycle(5, 0, 1, 2, 3, 4),
      Perm(Vector(0, 2, 4, 1, 3)) // (1 2 4 3) in 1-indexed
    ),
    maximalSupergroups = Vector(3, 4),
    compositionFactors = Vector(5, 2, 2)
  )

  /** A5 (T4): alternating group of order 60. */
  val a5: TGroup = TGroup(
    degree = 5, index = 4, name = "A5", order = 60, solvable = false,
    generators = Vector(
      Perm.cycle(5, 0, 1, 2, 3, 4),
      Perm.cycle(5, 0, 1, 2)
    ),
    maximalSupergroups = Vector(4)
  )

  /** S5 (T5): symmetric group of order 120. */
  val s5: TGroup = TGroup(
    degree = 5, index = 5, name = "S5", order = 120, solvable = false,
    generators = Vector(
      Perm.cycle(5, 0, 1, 2, 3, 4),
      Perm.transposition(5, 0, 1)
    )
  )

  /** All degree-5 transitive groups, ordered by index. */
  val degree5Groups: Vector[TGroup] = Vector(c5, d5, f20, a5, s5)

  /** Look up a degree-5 transitive group by name. */
  def byName5(name: String): Option[TGroup] =
    degree5Groups.find(_.name == name)

  // --- Degree 4 transitive groups ---

  /** C4 (T1): cyclic group of order 4. */
  val c4: TGroup = TGroup(
    degree = 4, index = 1, name = "C4", order = 4, solvable = true,
    generators = Vector(Perm.cycle(4, 0, 1, 2, 3))
  )

  /** V4 (T2): Klein four-group. */
  val v4: TGroup = TGroup(
    degree = 4, index = 2, name = "V4", order = 4, solvable = true,
    generators = Vector(
      Perm(Vector(1, 0, 3, 2)),
      Perm(Vector(2, 3, 0, 1))
    )
  )

  /** D4 (T3): dihedral group of order 8. */
  val d4: TGroup = TGroup(
    degree = 4, index = 3, name = "D4", order = 8, solvable = true,
    generators = Vector(
      Perm.cycle(4, 0, 1, 2, 3),
      Perm(Vector(0, 3, 2, 1))
    )
  )

  /** A4 (T4): alternating group of order 12. */
  val a4: TGroup = TGroup(
    degree = 4, index = 4, name = "A4", order = 12, solvable = true,
    generators = Vector(
      Perm.cycle(4, 0, 1, 2),
      Perm(Vector(1, 0, 3, 2))
    )
  )

  /** S4 (T5): symmetric group of order 24. */
  val s4: TGroup = TGroup(
    degree = 4, index = 5, name = "S4", order = 24, solvable = true,
    generators = Vector(
      Perm.cycle(4, 0, 1, 2, 3),
      Perm.transposition(4, 0, 1)
    )
  )

  /** All degree-4 transitive groups. */
  val degree4Groups: Vector[TGroup] = Vector(c4, v4, d4, a4, s4)

  // -----------------------------------------------------------------------
  // General prime-degree transitive groups (runtime computation)
  // -----------------------------------------------------------------------

  /** All transitive subgroups of S_n (up to conjugacy) for the given degree.
    *
    * For degree 5, returns the hard-coded database.
    * For other primes p >= 3, computes at runtime from AGL(1,p) structure.
    * For unsupported composite degrees, returns empty.
    */
  def transGroupsOfDegree(n: Int): Vector[TGroup] =
    if n == 5 then degree5Groups
    else if n >= 3 && PrimeFactors.isPrime(n) then transGroupsOfPrimeRT(n)
    else Vector.empty

  /** Find transitive group(s) of the given degree and order. */
  def transGroupByOrder(deg: Int, ord: Long): Vector[TGroup] =
    transGroupsOfDegree(deg).filter(_.order == ord)

  /** Modular exponentiation: base^exp mod m. */
  def modExp(base: Long, exp: Long, m: Long): Long =
    if exp == 0 then 1L
    else if exp % 2 == 0 then
      val half = modExp(base, exp / 2, m)
      (half * half) % m
    else (base % m * modExp(base, exp - 1, m)) % m

  /** Primitive root modulo a prime p. */
  def primitiveRootP(p: Long): Long =
    val phi = p - 1
    val factors = primeFactorsFlat(phi).distinct
    (2L until p).find { g =>
      factors.forall(q => modExp(g, phi / q, p) != 1L)
    }.getOrElse(2L)

  /** Sorted positive divisors of n. */
  def divisors(n: Long): Vector[Long] =
    val isqrt = math.sqrt(n.toDouble).toLong
    val small = (1L to isqrt).filter(d => n % d == 0)
    val all = small.flatMap(d => if d * d == n then Vector(d) else Vector(d, n / d))
    all.distinct.sorted.toVector

  /** Prime factors with multiplicity (flat list). */
  private def primeFactorsFlat(n: Long): Vector[Long] =
    if n <= 1 then Vector.empty
    else
      val pos = Positive.unsafe(n.toInt)
      PrimeFactors.factorise(pos).flatMap { case (p, e) =>
        Vector.fill(e)(p.toLong)
      }

  /** Build all transitive groups for prime p at runtime.
    *
    * Solvable groups: Z/p ⋊ H for each divisor d of p-1.
    * Non-solvable: A_p, S_p.
    * Maximal supergroup relationships are computed from divisibility.
    */
  private def transGroupsOfPrimeRT(p: Int): Vector[TGroup] =
    val pL = p.toLong
    val g = primitiveRootP(pL)
    val ds = divisors(pL - 1)

    // Build one solvable group per divisor of p-1
    val solvableGroups = ds.map { d =>
      val trans = Perm(Vector.tabulate(p)(i => (i + 1) % p))
      val scaleFactor = modExp(g, (pL - 1) / d, pL)
      val scale = Perm(Vector.tabulate(p)(i => ((scaleFactor * i) % pL).toInt))
      val gens = if d == 1 then Vector(trans) else Vector(trans, scale)
      val gName =
        if d == 1 then s"Z$p"
        else if d == 2 then s"D$p"
        else if d == pL - 1 then s"AGL(1,$p)"
        else s"Z$p:Z$d"
      val cFactors = primeFactorsFlat(d).map(_.toInt) :+ p
      TGroup(
        degree = p, index = 0, name = gName, order = pL * d,
        solvable = true, generators = gens,
        compositionFactors = cFactors
      )
    }

    // A_p
    val apOrder = factorial(pL) / 2
    val ap = TGroup(
      degree = p, index = 0, name = s"A$p", order = apOrder,
      solvable = p < 5,
      generators = Vector(
        Perm.cycle(p, (0 until p)*),
        Perm.cycle(p, 0, 1, 2)
      )
    )

    // S_p
    val spOrder = factorial(pL)
    val sp = TGroup(
      degree = p, index = 0, name = s"S$p", order = spOrder,
      solvable = p < 4,
      generators = Vector(
        Perm.cycle(p, (0 until p)*),
        Perm.transposition(p, 0, 1)
      )
    )

    val allGroups = (solvableGroups :+ ap :+ sp).sortBy(_.order)

    // Assign indices and maximal supergroups
    val indexed = allGroups.zipWithIndex
    indexed.map { case (tg, myIdx) =>
      val myOrd = tg.order
      val cands = indexed.filter { case (cg, i) =>
        i != myIdx && cg.order > myOrd && cg.order % myOrd == 0
      }
      val maxSupers = cands.filter { case (_, superIdx) =>
        val superOrd = allGroups(superIdx).order
        !cands.exists { case (midG, _) =>
          val midOrd = midG.order
          midOrd > myOrd && midOrd < superOrd &&
            superOrd % midOrd == 0 && midOrd % myOrd == 0
        }
      }.map(_._2)
      tg.copy(index = myIdx + 1, maximalSupergroups = maxSupers)
    }

  /** Compute n! (for moderate n). */
  private def factorial(n: Long): Long =
    (1L to n).product

  // -----------------------------------------------------------------------
  // Composition series
  // -----------------------------------------------------------------------

  /** For a solvable transitive group, return the composition series as a
    * list of generating sets, descending from G to the trivial group {1}.
    *
    * Each consecutive quotient G_i / G_{i+1} is cyclic of prime order
    * (matching the composition factors). This chain drives radical tower
    * descent.
    *
    * Returns None for non-solvable groups.
    */
  def compositionSeries(tg: TGroup): Option[Vector[Vector[Perm]]] =
    if !tg.solvable then None
    else tg.name match
      // Fast path: degree-5 hard-coded
      case "C5" => Some(Vector(tg.generators, Vector.empty))
      case "D5" => Some(Vector(
        tg.generators,
        Vector(Perm.cycle(5, 0, 1, 2, 3, 4)),
        Vector.empty
      ))
      case "F20" => Some(Vector(
        tg.generators,
        Vector(Perm.cycle(5, 0, 1, 2, 3, 4), Perm(Vector(0, 4, 3, 2, 1))),
        Vector(Perm.cycle(5, 0, 1, 2, 3, 4)),
        Vector.empty
      ))
      case _ if PrimeFactors.isPrime(tg.degree) =>
        compositionSeriesPrime(tg)
      case _ => None

  /** Composition series for a solvable affine subgroup of S_p.
    *
    * The group has the form Z/p ⋊ H where H is cyclic of order d.
    * The series descends through subgroups of H by removing one
    * prime factor at a time, then drops to {1}.
    */
  private def compositionSeriesPrime(tg: TGroup): Option[Vector[Vector[Perm]]] =
    val p = tg.degree.toLong
    val n = tg.degree
    val g = primitiveRootP(p)
    val d = tg.order / p
    val dFactors = primeFactorsFlat(d)
    // Chain of divisors: d, d/q1, d/(q1*q2), ..., 1
    val dChain = dFactors.scanLeft(d)((acc, q) => acc / q)
    val trans = Perm.cycle(n, (0 until n)*)

    def mkGens(dPrime: Long): Vector[Perm] =
      if dPrime <= 1 then Vector(trans)
      else
        val sf = modExp(g, (p - 1) / dPrime, p)
        val scale = Perm(Vector.tabulate(n)(i => ((sf * i) % p).toInt))
        Vector(trans, scale)

    Some(dChain.map(mkGens) :+ Vector.empty)
